### PR TITLE
Use monitor notifications to wake up from sleep in the main keeper loop.

### PR DESCRIPTION
**Pull request by DimCitus**
This allows to shorten considerably our reaction time when the monitor is
orchestrating our group, and also means we can display information about the
other nodes in the cluster in each node logs, which makes debugging easier.

### CODE REVIEW

File: src/bin/pg_autoctl/monitor.c [hunk 1]
```diff
@@ -3075,6 +3125,128 @@ monitor_get_notifications(Monitor *monitor)
 }
 
 
+/*
+ * monitor_wait_for_state_change waits for timeout milliseconds or until we
+ * receive a notification for a state change concerning the given nodeId,
```
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:3130]: There's no `nodeId` parameter.
```suggestion
 * receive a notification for a state change concerning the given groupId,
```

**Review by JelteF — CHANGES_REQUESTED** [review]
I did some QA and it seems the new code is not always working as expected:
```
    Time |   Name |  Node |           Host:Port |       Current State |      Assigned State
---------+--------+-------+---------------------+---------------------+--------------------
14:40:42 |  node2 |     2 | 172.30.151.233:5502 |             primary |        join_primary
14:40:42 | node_1 |     1 | 172.30.151.233:5501 |           secondary |    wait_maintenance
14:40:42 | node_1 |     1 | 172.30.151.233:5501 |    wait_maintenance |    wait_maintenance
14:40:42 |  node2 |     2 | 172.30.151.233:5502 |        join_primary |        join_primary
14:40:52 | node_1 |     1 | 172.30.151.233:5501 |    wait_maintenance |         maintenance
14:40:52 |  node2 |     2 | 172.30.151.233:5502 |        join_primary |             primary
14:40:52 | node_1 |     1 | 172.30.151.233:5501 |         maintenance |         maintenance
```

There's now a 10 second delay between reaching `join_primary` on `node2` and asigning `maintenance` state to `node_1`. I assume this is related to the increased 10 second keeper timeout. I think either the root cause should be fixed, or the keeper timeout should be decreased again.



Also, not sure if related to this PR, but there's something strange (maybe expected though) with waiting `disable maintenance` if it is done right after `enable maintenance`:
```
pg_autoctl enable maintenance --pgdata  node1 && pg_autoctl disable maintenance --pgdata node1
14:44:50 1976 INFO  Listening monitor notifications about state changes in formation "default" and group 0
14:44:50 1976 INFO  Following table displays times when notifications are received
    Time |   Name |  Node |           Host:Port |       Current State |      Assigned State
---------+--------+-------+---------------------+---------------------+--------------------
14:44:50 |  node2 |     2 | 172.30.151.233:5502 |             primary |        join_primary
14:44:50 | node_1 |     1 | 172.30.151.233:5501 |           secondary |    wait_maintenance
14:44:50 | node_1 |     1 | 172.30.151.233:5501 |    wait_maintenance |    wait_maintenance
14:44:50 |  node2 |     2 | 172.30.151.233:5502 |        join_primary |        join_primary
14:45:00 | node_1 |     1 | 172.30.151.233:5501 |    wait_maintenance |         maintenance
14:45:00 |  node2 |     2 | 172.30.151.233:5502 |        join_primary |             primary
14:45:00 | node_1 |     1 | 172.30.151.233:5501 |         maintenance |         maintenance
14:45:00 2469 INFO  Listening monitor notifications about state changes in formation "default" and group 0
14:45:00 2469 INFO  Following table displays times when notifications are received
    Time |   Name |  Node |           Host:Port |       Current State |      Assigned State
---------+--------+-------+---------------------+---------------------+--------------------
14:45:01 | node_1 |     1 | 172.30.151.233:5501 |         maintenance |          catchingup
14:45:01 |  node2 |     2 | 172.30.151.233:5502 |             primary |             primary
```

If in between enable and disable we wait a bit this is the disable output:
```
14:46:40 7473 INFO  Listening monitor notifications about state changes in formation "default" and group 0
14:46:40 7473 INFO  Following table displays times when notifications are received
    Time |   Name |  Node |           Host:Port |       Current State |      Assigned State
---------+--------+-------+---------------------+---------------------+--------------------
14:46:41 | node_1 |     1 | 172.30.151.233:5501 |         maintenance |          catchingup
14:46:41 | node_1 |     1 | 172.30.151.233:5501 |          catchingup |          catchingup
14:46:41 | node_1 |     1 | 172.30.151.233:5501 |          catchingup |           secondary
14:46:41 |  node2 |     2 | 172.30.151.233:5502 |             primary |             primary
```

**Comment by JelteF** [comment]
Also with 2 standbys during perform failover the logs now show these errors:
```
14:34:08 28678 INFO  Ensuring HBA rules for node 1 (172.30.151.233:5501)
14:34:08 28678 INFO  Ensuring HBA rules for node 3 (172.30.151.233:5503)
14:34:08 28678 INFO  Transition complete: current state is now "wait_primary"
14:34:08 28678 INFO  New state for node 3 "node3" (172.30.151.233:5503): report_lsn ➜ secondary
14:34:08 28678 INFO  New state for node 2 "node2" (172.30.151.233:5502): wait_primary ➜ primary
14:34:08 28678 INFO  Monitor assigned new state "primary"
14:34:08 28678 INFO  FSM transition from "wait_primary" to "primary": A healthy secondary appeared
14:34:08 28678 INFO  Setting synchronous_standby_names to 'ANY 1 (pgautofailover_standby_1, pgautofailover_standby_3)'
14:34:08 28678 WARN  Failed to set the standby Target LSN because we don't have a quorum candidate yet
14:34:08 28678 ERROR Failed to transition from state "wait_primary" to state "primary", see above.
14:34:08 28678 ERROR Failed to transition to state "primary", retrying... 
```

**Comment by DimCitus** [comment]
> There's now a 10 second delay between reaching `join_primary` on `node2` and asigning `maintenance` state to `node_1`. I assume this is related to the increased 10 second keeper timeout. I think either the root cause should be fixed, or the keeper timeout should be decreased again.

I believe this was a case of cache invalidation problem where we would forget to set the `notificationReceived` tracker boolean to true when we did receive a notification. In my local testing, that's fixed now.

**Comment by JelteF** [comment]
Enabling maintenance can still take 10 seconds now:
```
16:26:54 20298 INFO  Listening monitor notifications about state changes in formation "default" and group 0
16:26:54 20298 INFO  Following table displays times when notifications are received
    Time |   Name |  Node |           Host:Port |       Current State |      Assigned State
---------+--------+-------+---------------------+---------------------+--------------------
16:26:54 | node_1 |     1 | 172.28.149.103:5501 |             primary |        wait_primary
16:26:54 | node_2 |     2 | 172.28.149.103:5502 |           secondary |    wait_maintenance
16:26:54 | node_2 |     2 | 172.28.149.103:5502 |    wait_maintenance |    wait_maintenance
16:26:54 | node_1 |     1 | 172.28.149.103:5501 |        wait_primary |        wait_primary
16:27:05 | node_2 |     2 | 172.28.149.103:5502 |    wait_maintenance |         maintenance
16:27:05 | node_2 |     2 | 172.28.149.103:5502 |         maintenance |         maintenance
```

File: src/bin/pg_autoctl/primary_standby.c [hunk 2]
```diff
@@ -733,6 +733,11 @@ standby_init_replication_source(LocalPostgresServer *postgres,
 
 	if (primaryNode != NULL)
 	{
+		upstream->primaryNode.nodeId = primaryNode->nodeId;
+
+		strlcpy(upstream->primaryNode.name,
+				primaryNode->name, _POSIX_HOST_NAME_MAX);
```
Comment (JelteF) [inline src/bin/pg_autoctl/primary_standby.c:739]: Does `name` have a maximum of `_POSIX_HOST_NAME_MAX`? I would expect `MAX_BUFSIZE`.

**Review by JelteF — APPROVED** [review]
I'm good with merging this if these final two comments are addressed.

File: src/bin/pg_autoctl/primary_standby.c (hunk 2 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/primary_standby.c:739]: Indeed, our node name has the same length as a hostname:

``` c
	char name[_POSIX_HOST_NAME_MAX];
	char hostname[_POSIX_HOST_NAME_MAX];
```

File: src/bin/pg_autoctl/defaults.h [hunk 3]
```diff
@@ -70,7 +70,8 @@
 #define PREPARE_PROMOTION_CATCHUP_TIMEOUT 30
 #define PREPARE_PROMOTION_WALRECEIVER_TIMEOUT 5
 
-#define PG_AUTOCTL_KEEPER_SLEEP_TIME 1
+#define PG_AUTOCTL_KEEPER_SLEEP_TIME 10     /* seconds */
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/defaults.h]: Changed back to 1s.

**Pull request merged**